go-controller/pkg/network-controller-manager/network_attach_def_controller.go

```diff
@@ -0,0 +1,526 @@
+			nadController.perNADNetConfInfo.LoadOrStore(nadName, &nadNetConfInfo{NetConfInfo: netConfInfo, netName: netName})
+			err = nadController.addNADToController(ncm, nadName, nInfo, netConfInfo, doStart)
+			if err != nil {
+				klog.Errorf("Failed to add net-attach-def %s to network %s: %v", nadName, netName, err)
+				nadController.perNADNetConfInfo.Delete(nadName)
+				return err
+			}
+			return nil
+		}
+		return nil
+	})
+}
+
+// DeleteNetAttachDef deletes the given NAD from the associated controller. It delete the controller if this
+// is the last NAD of the network
+func (nadController *netAttachDefinitionController) DeleteNetAttachDef(netAttachDefName string) error {
+	klog.Infof("Delete net-attach-def %s", netAttachDefName)
+	return nadController.perNADNetConfInfo.DoWithLock(netAttachDefName, func(nadName string) error {
+		existingNadNetConfInfo, found := nadController.perNADNetConfInfo.Load(nadName)
+		if !found {
+			klog.V(5).Infof("net-attach-def %s not found for removal", nadName)
+			return nil
+		}
+		err := nadController.deleteNADFromController(existingNadNetConfInfo.netName, nadName)
+		if err != nil {
+			klog.Errorf("Failed to delete net-attach-def %s from network %s: %v", nadName, existingNadNetConfInfo.netName, err)
+			return err
+		}
+		nadController.perNADNetConfInfo.Delete(nadName)
+		return nil
+	})
+}
+
+func (nadController *netAttachDefinitionController) addNADToController(ncm NetworkControllerManager, nadName string,
+	nInfo util.NetInfo, netConfInfo util.NetConfInfo, doStart bool) (err error) {
+	var oc NetworkController
+	var nadExists, isStarted bool
+
+	netName := nInfo.GetNetworkName()
+	klog.V(5).Infof("Add net-attach-def %s to network %s", nadName, netName)
+	return nadController.perNetworkNADInfo.DoWithLock(netName, func(networkName string) error {
+		nni, loaded := nadController.perNetworkNADInfo.LoadOrStore(networkName, &networkNADInfo{
+			nadNames:  map[string]bool{},
+			oc:        nil,
+			isStarted: false,
+		})
+		if !loaded {
+			defer func() {
+				if err != nil {
+					nadController.perNetworkNADInfo.Delete(networkName)
+				}
+			}()
+			// first NAD for this network, create controller
+			klog.V(5).Infof("First net-attach-def %s of network %s added, create network controller", nadName, networkName)
+			oc, err = ncm.NewNetworkController(nInfo, netConfInfo)
+			if err != nil {
+				return fmt.Errorf("failed to create network controller for network %s: %v", networkName, err)
+			}
+			nni.oc = oc
+		} else {
+			klog.V(5).Infof("net-attach-def %s added to existing network %s", nadName, networkName)
+			// controller of this network already exists
+			oc = nni.oc
+			isStarted = nni.isStarted
+			_, nadExists = nni.nadNames[nadName]
+
+			if !oc.CompareNetConf(netConfInfo) {
+				if nadExists {
+					// this should not happen, continue to start the existing controller if requested
+					klog.Warningf("net-attach-def %s netconf spec changed, should not happen", networkName)
```

Comment:
@cathy-zhou what did you decide to do here? As I said, if it should not happen I would prefer to return an error. If there is specific cases when this can happen, can we address that in the comment?